lib/sensors/GPS.cpp

```diff
@@ -0,0 +1,483 @@
+//Settings: 	use COM_TYPE_ constants
+void GPS::setPortOutput(uint8_t portID, uint8_t outStreamSettings) {
+	//Get the current config values for this port ID, loads payloadCfg array
+	getPortSettings(portID);
+	m_packetCfg.cls = UBX_CLASS_CFG;
+	m_packetCfg.id = UBX_CFG_PRT;
+	m_packetCfg.len = 20;
+	m_packetCfg.startingSpot = 0;
+	//payloadCfg is now loaded with current bytes. Change only the ones we need to
+	payloadCfg[14] = outStreamSettings;
+	sendSerialCommand(&m_packetCfg);
+}
+
+//Configure a given port to accept UBX, NMEA, RTCM3 or a combination thereof
+//Port:     	use COM_PORT_ constants
+//Settings: 	use COM_TYPE_ constants
+void GPS::setPortInput(uint8_t portID, uint8_t inStreamSettings) {
+	//Get the current config values for this port ID, loads payloadCfg array
+	m_packetCfg.cls = UBX_CLASS_CFG;
+	m_packetCfg.id = UBX_CFG_PRT;
+	m_packetCfg.len = 20;
+	m_packetCfg.startingSpot = 0;
+	//payloadCfg is now loaded with current bytes. Change only the ones we need to
+	payloadCfg[12] = inStreamSettings;
+	sendSerialCommand(&m_packetCfg);
+}
+
+void GPS::setNavigationFrequency(uint8_t navFreq) {
+	m_packetCfg.cls = UBX_CLASS_CFG;
+	m_packetCfg.id = UBX_CFG_RATE;
+	m_packetCfg.len = 0;
+	m_packetCfg.startingSpot = 0;
+
+	//This will load the payloadCfg array with current settings of the given register
+	sendSerialCommand(&m_packetCfg);
+
+	uint16_t measurementRate = 1000 / navFreq;
+
+	//payloadCfg is now loaded with current bytes. Change only the ones we need to
+	payloadCfg[0] = measurementRate & 0xFF; //measRate LSB
+	payloadCfg[1] = measurementRate >> 8;   //measRate MSB
+
+	sendSerialCommand(&m_packetCfg);
+}
+
+void GPS::configureMessage(uint8_t msgClass, uint8_t msgID, uint8_t portID, uint8_t sendRate) {
+	m_packetCfg.cls = UBX_CLASS_CFG;
+	m_packetCfg.id = UBX_CFG_MSG;
+	m_packetCfg.len = 8;
+	m_packetCfg.startingSpot = 0;
+
+	//Clear packet payload
+	for (int x = 0; x < m_packetCfg.len; x++) {
+		m_packetCfg.payload[x] = 0;
+	}
+	m_packetCfg.payload[0] = msgClass;
+	m_packetCfg.payload[1] = msgID;
+	//Send rate is relative to the event a message is registered on.
+	// For example, if the rate of a navigation message is set to 2, the message is sent every 2nd navigation solution.
+	m_packetCfg.payload[2 + portID] = sendRate;
+
+	sendSerialCommand(&m_packetCfg);
+}
+
+void GPS::saveConfiguration()
+// saved config is automatically loaded on startup
+// will allow configuration to persist through module resets
+// where initial configuration would not necessarily run again
+{
+	m_packetCfg.cls = UBX_CLASS_CFG;
+	m_packetCfg.id = UBX_CFG_CFG;
+	m_packetCfg.len = 12;
+	m_packetCfg.startingSpot = 0;
+
+	//Clear packet payload
+	//TODO: make sure replacing uint8_t with int didn't break anything here
+	for (int x = 0; x < m_packetCfg.len; x++) {
+		m_packetCfg.payload[x] = 0;
+	}
+
+	m_packetCfg.payload[4] = 0xFF; //Setting saveMask bits
+	m_packetCfg.payload[5] = 0xFF;
+
+	sendSerialCommand(&m_packetCfg);
+}
```

Comment:
This allows to save config of the calibration settings?

Reply:
This saves the config in the receiver's non-volatile memory